src/managers/backup.py

```diff
@@ -84,6 +93,87 @@ def create_bucket(self, s3_parameters: S3ConnectionInfo) -> bool:
 
         return True
 
-    def write_test_string(self) -> None:
-        """Write content in the object storage."""
-        self.bucket.put_object(Key="test_file.txt", Body=b"test string")
+    def create_backup(self) -> BackupMetadata:
+        """Create a snapshot with ZooKeeper admin server and stream it to the object storage."""
+        zk_user = "super"
+        zk_pwd = self.state.cluster.internal_user_credentials.get("super", "")
+        date = datetime.now()
+        snapshot_name = f"{date:%Y-%m-%dT%H:%M:%SZ}"
+        snapshot_path = f"{snapshot_name}/snapshot"
+
+        # It is very likely that the file is fully loaded in memory, because the file-like interface is
+        # not seekable, and I have a strong suspicion that boto uses this to figure out if it can
+        # upload in one go or need to use a multipart request.
+        # We cannot be sure because finding this information in boto code base is time consuming.
+        # If this ever become an issue, we can find a workaround by using the 'content-length' header from
+        # the response. Or write to a temp file as a last resort.
+        with httpx.stream(
+            "GET",
+            f"http://localhost:{ADMIN_SERVER_PORT}/commands/snapshot?streaming=true",
+            headers={"Authorization": f"digest {zk_user}:{zk_pwd}"},
+        ) as response:
+
+            response_headers = response.headers
+            quorum_leader_zxid = int(response_headers["last_zxid"], base=16)
+            metadata: BackupMetadata = {
+                "id": snapshot_name,
```

Comment:
**question:** Curiosity, are we expecting this `id` to be incremental? i.e Backup 1 --> Backup 2 etc?

Reply:
I realise that a collision is possible. `id` as per what other charms do is the datetime serialised. So it's not impossible, though unlikely in normal conditions, that a snapshot already exists with the same id. I do not have a good answer as to how we should handle this case

Reply:
I think my original question was a misunderstanding on the `id`. Could we ID our backups using the zxid? It's a number after all, and DEFINITELY should be incremental.

Reply:
No, for consistency with other charms we need to use the same naming convention for the snapshots

Reply:
I get that, but I'll share my reasoning:
- A serialised datetime corresponds to a unique point in time, from which a backup was created
- Zxid is a 64-bit number, made up of 32-bits for `epoch` and 32-bits for `count`. A new leader == new `epoch`, and a new proposal increments the `count`
- Because of this atomic log, it sort of **is** a measure of 'time', just 'ZooKeeper time'
- If we use it, we're still using it the same as 'the point in which a backup was created', just using a different relative measure for time

Reply:
But, it really doesn't matter, as the information is already in `log-sequence-number` anyway 👍🏾